// js-waku/build/main/lib/discovery/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPseudoRandomSubset = exports.ENRTree = exports.DnsOverHttps = exports.DnsNodeDiscovery = exports.Bootstrap = exports.getNodesFromHostedJson = void 0;
const utils_1 = require("libp2p-gossipsub/src/utils");
var hosted_json_1 = require("./hosted_json");
Object.defineProperty(exports, "getNodesFromHostedJson", { enumerable: true, get: function () { return hosted_json_1.getNodesFromHostedJson; } });
var bootstrap_1 = require("./bootstrap");
Object.defineProperty(exports, "Bootstrap", { enumerable: true, get: function () { return bootstrap_1.Bootstrap; } });
var dns_1 = require("./dns");
Object.defineProperty(exports, "DnsNodeDiscovery", { enumerable: true, get: function () { return dns_1.DnsNodeDiscovery; } });
var dns_over_https_1 = require("./dns_over_https");
Object.defineProperty(exports, "DnsOverHttps", { enumerable: true, get: function () { return dns_over_https_1.DnsOverHttps; } });
var enrtree_1 = require("./enrtree");
Object.defineProperty(exports, "ENRTree", { enumerable: true, get: function () { return enrtree_1.ENRTree; } });
function getPseudoRandomSubset(values, wantedNumber) {
    if (values.length <= wantedNumber) {
        return values;
    }
    return utils_1.shuffle(values).slice(0, wantedNumber);
}
exports.getPseudoRandomSubset = getPseudoRandomSubset;
